src/runtime/uv/tcp.cpp

```diff
@@ -0,0 +1,611 @@
+    connect_data->socket = socket;
+    uv_connect->data = connect_data;
+
+    // The event loop owns the socket.
+    lean_inc(socket);
+    lean_inc(promise);
+
+    event_loop_lock(&global_ev);
+
+    int result = uv_tcp_connect(uv_connect, tcp_socket->m_uv_tcp, (sockaddr*)&addr_ptr, [](uv_connect_t* req, int status) {
+        tcp_connect_data* tup = (tcp_connect_data*) req->data;
+        lean_promise_resolve_with_code(status, tup->promise);
+
+        // The event loop does not own the object anymore.
+        lean_dec(tup->socket);
+        lean_dec(tup->promise);
+
+        free(req->data);
+        free(req);
+    });
+
+    event_loop_unlock(&global_ev);
+
+    if (result < 0) {
+        lean_dec(promise); // The structure does not own it.
+        lean_dec(promise); // We are not going to return it.
+        lean_dec(socket);
+
+        free(uv_connect->data);
+        free(uv_connect);
+
+        return lean_io_result_mk_error(lean_decode_uv_error(result, nullptr));
+    }
+
+    return lean_io_result_mk_ok(promise);
+}
+
+/* Std.Internal.UV.TCP.Socket.send (socket : @& Socket) (data : ByteArray) : IO (IO.Promise (Except IO.Error Unit)) */
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_send(b_obj_arg socket, obj_arg data) {
+    lean_uv_tcp_socket_object* tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    size_t data_len = lean_sarray_size(data);
+    char* data_str = (char*)lean_sarray_cptr(data);
+
+    uv_buf_t buf = uv_buf_init(data_str, data_len);
+
+    lean_object* promise = lean_promise_new();
+    mark_mt(promise);
+
+    uv_write_t* write_uv = (uv_write_t*)malloc(sizeof(uv_write_t));
+    write_uv->data = (tcp_send_data*)malloc(sizeof(tcp_send_data));
+
+    tcp_send_data* send_data = (tcp_send_data*)write_uv->data;
+    send_data->promise = promise;
+    send_data->data = data;
+    send_data->socket = socket;
+
+    // These objects are going to enter the loop and be owned by it
+    lean_inc(promise);
+    lean_inc(socket);
+
+    event_loop_lock(&global_ev);
+
+    int result = uv_write(write_uv, (uv_stream_t*)tcp_socket->m_uv_tcp, &buf, 1, [](uv_write_t* req, int status) {
+        tcp_send_data* tup = (tcp_send_data*) req->data;
+
+        lean_promise_resolve_with_code(status, tup->promise);
+
+        lean_dec(tup->promise);
+        lean_dec(tup->data);
+        lean_dec(tup->socket);
+
+        free(req->data);
+        free(req);
+    });
+
+    event_loop_unlock(&global_ev);
+
+    if (result < 0) {
+        lean_dec(promise); // The structure does not own it.
+        lean_dec(promise); // We are not going to return it.
+        lean_dec(socket);
+
+        lean_dec(data);
+
+        free(write_uv->data);
+        free(write_uv);
+
+        return lean_io_result_mk_error(lean_decode_uv_error(result, nullptr));
+    }
+
+    return lean_io_result_mk_ok(promise);
+}
+
+/* Std.Internal.UV.TCP.Socket.recv? (socket : Socket) (size : UInt64) : IO (IO.Promise (Except IO.Error (Option ByteArray))) */
```

Comment:
This lean type is still not quite correct:
```suggestion
/* Std.Internal.UV.TCP.Socket.recv? (socket : @& Socket) (size : UInt64) : IO (IO.Promise (Except IO.Error (Option ByteArray))) */
```